### DISCUSSION TITLE
GetMappedRange doesn't feel very intuitive and requires a complex tracking logic in the backend

### DISCUSSION THREAD

**Discussion by prollin** [gpuweb/gpuweb#4901] [Ideas/Feedback]
WebGPU is overall a great API and is very close to what a "sane" GPU API abstraction should look like.
While trying to match some of it in my own lightweight abstraction, I found that implementing `GetMappedRange` (tracking each range internally) is both cumbersome and not really useful as a feature.
I remember the early days of WebGPU where we had `mapReadAsync` and `mapWriteAsync` that returned the mapped data as part of the promise which felt nice and allowed to protect against any unwanted access to the mapped data.

I most likely missed some meeting minutes somewhere that already cover this, but why did WebGPU switched to allow `GetMappedRange` to return sub ranges of the already mapped data? Is there some browser specific issue that this alleviate?

While it is most likely too late to fix at this point, going back to `mapReadAsync/mapWriteAsync` or simplify `GetMappedRange()` to simply return the currently mapped range (in full) would make for a less error prone experience and simpler implementation on the backend. This is what I do in my abstraction, and I haven't found a use case where I need a subrange so far.

**Comment by kainino0x** [comment]
getMappedRange lets the browser/implementation know which ranges could possibly have changed and need to be invalidated/flushed. Since you can only have one mapAsync-mapping at a time, you might for example need to map a very large range, even when you only need to update some smaller sections of data. getMappedRange lets you update those smaller sections without forcing the entire mapped region to be re-uploaded.

(Note if the updates needed are small, then writeBuffer is a good alternative in many cases.)

We actually have a recent issue open to clarify this in the spec: #4805

**Reply by prollin** [reply]
If I understand correctly, browser safety requirement around memory makes it so the user can't get direct access to GPU memory and when a buffer is mapped, the WebGPU implementation has to manage a copy that is then flushed to GPU memory when the buffer is unmapped (feel free to poke holes in that assumption, I am in no way a web expert).
If it is about optimizing update to small portions of a buffer, I don't see how this can be more efficient (both in term of performance and memory overhead) than a proper writeBuffer implementation (which will have to manage a copy as well) as you also suggest.

My point here is that from an API surface standpoint, getMappedRange feels like an unnecessary feature that is prone to error (offset) and makes implementations more convoluted. writeBuffer is easier to optimize internally and works with all buffers regardless of usage.
For reading, it is surely not useful at all and the older mapReadAsync seemed much easier for anyone to understand and work with.

**Reply by kainino0x** [reply]
Mapping doesn't necessarily involve a copy, it's possible to expose the mapping directly from the GPU driver (though no browsers currently do this AFAIK). If that's the case, then flushing ranges might not involve a copy, but instead something like a cache-flush on the CPU side and cache-invalidation on the GPU side.

writeBuffer additionally requires an extra copy because it pulls data from a source ArrayBuffer, rather than providing one to write into. This is why we recommend writeBuffer mostly for smaller uploads. In principle, at least for some upload patterns, maintaining your own staging buffer and mapping it can be more efficient.
